File: app/src/main/java/com/spidroid/starry/models/StoryModel.java
package com.spidroid.starry.models;

import com.google.firebase.firestore.ServerTimestamp;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.UUID; // لإنشاء معرّف فريد للقصة

public class StoryModel {

    private String storyId;
    private String userId; // معرف المستخدم الذي أنشأ القصة
    private String mediaUrl; // رابط الصورة/الفيديو للقصة
    private String thumbnailUrl; // رابط الصورة المصغرة للفيديو (إذا كانت القصة فيديو)
    private String mediaType; // "image" أو "video"
    private long duration; // مدة القصة بالمللي ثانية (خاصة بالفيديو)
    @ServerTimestamp
    private Date createdAt; // وقت إنشاء القصة
    private Date expiresAt; // وقت انتهاء صلاحية القصة (بعد 24 ساعة من الإنشاء)
    private Map<String, Boolean> viewers = new HashMap<>(); // معرفات المستخدمين الذين شاهدوا القصة

    // أنواع الوسائط
    public static final String MEDIA_TYPE_IMAGE = "image";
    public static final String MEDIA_TYPE_VIDEO = "video";

    public StoryModel() {
        // مطلوب لدوال Firebase
    }

    public StoryModel(String userId, String mediaUrl, String mediaType, long duration, String thumbnailUrl) {
        this.storyId = UUID.randomUUID().toString(); // إنشاء معرّف فريد للقصة
        this.userId = userId;
        this.mediaUrl = mediaUrl;
        this.mediaType = mediaType;
        this.duration = duration;
        this.thumbnailUrl = thumbnailUrl;
        this.createdAt = new Date();
        this.expiresAt = new Date(System.currentTimeMillis() + 24 * 60 * 60 * 1000); // تنتهي بعد 24 ساعة
    }

    // Getters
    public String getStoryId() {
        return storyId;
    }

    public String getUserId() {
        return userId;
    }

    public String getMediaUrl() {
        return mediaUrl;
    }

    public String getThumbnailUrl() {
        return thumbnailUrl;
    }

    public String getMediaType() {
        return mediaType;
    }

    public long getDuration() {
        return duration;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public Date getExpiresAt() {
        return expiresAt;
    }

    public Map<String, Boolean> getViewers() {
        return viewers;
    }

    // Setters
    public void setStoryId(String storyId) {
        this.storyId = storyId;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    public void setMediaUrl(String mediaUrl) {
        this.mediaUrl = mediaUrl;
    }

    public void setThumbnailUrl(String thumbnailUrl) {
        this.thumbnailUrl = thumbnailUrl;
    }

    public void setMediaType(String mediaType) {
        this.mediaType = mediaType;
    }

    public void setDuration(long duration) {
        this.duration = duration;
    }

    public void setCreatedAt(Date createdAt) {
        this.createdAt = createdAt;
    }

    public void setExpiresAt(Date expiresAt) {
        this.expiresAt = expiresAt;
    }

    public void setViewers(Map<String, Boolean> viewers) {
        this.viewers = viewers;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        StoryModel that = (StoryModel) o;
        return Objects.equals(storyId, that.storyId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(storyId);
    }
}

File: app/src/main/java/com/spidroid/starry/models/StoryModel.kt
// hmze123/sambok/sambok-main/app/src/main/java/com/spidroid/starry/models/StoryModel.kt
package com.spidroid.starry.models

import com.google.firebase.firestore.Exclude
import com.google.firebase.firestore.ServerTimestamp
import java.util.Date
import java.util.Objects
import java.util.UUID

class StoryModel {
    // --- Getters and Setters ---
    // --- الحقول الأساسية ---
    var storyId: String? = null // ✨ تم تغيير هذا ليصبح قابلاً للـ null
    var userId: String? = null
    var mediaUrl: String? = null
    var thumbnailUrl: String? = null
    var mediaType: String? = null
    var duration: Long = 0

    @ServerTimestamp
    var createdAt: Date? = null
    var expiresAt: Date? = null
    var viewers: MutableMap<String, Boolean> = mutableMapOf() // ✨ تم تغيير التهيئة إلى mutableMapOf() وإزالة Nullability في القيم


    // --- حقول بيانات المؤلف (لتحسين الأداء) ---
    @Exclude // ✨ تم إزالة @get:Exclude المكرر
    var authorUsername: String? = null

    @Exclude // ✨ تم إزالة @get:Exclude المكرر
    var authorDisplayName: String? = null

    @Exclude // ✨ تم إزالة @get:Exclude المكرر
    var authorAvatarUrl: String? = null

    // ** الدوال الجديدة التي تم إضافتها **
    @Exclude // ✨ تم إزالة @get:Exclude المكرر
    var isAuthorVerified: Boolean = false // ** تم إضافة هذا الحقل **


    // --- المُنشئات ---
    constructor()

    constructor(
        userId: String?,
        mediaUrl: String?,
        mediaType: String?,
        duration: Long,
        thumbnailUrl: String?
    ) {
        // ✨ لا يتم تعيين storyId هنا، بل يتم تركه لـ Firestore ليقوم بتعيينه
        this.userId = userId
        this.mediaUrl = mediaUrl
        this.mediaType = mediaType
        this.duration = duration
        this.thumbnailUrl = thumbnailUrl
        this.createdAt = Date()
        this.expiresAt = Date(System.currentTimeMillis() + 24 * 60 * 60 * 1000)
    }

    // --- دوال مساعدة ---
    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as StoryModel
        return storyId == that.storyId
    }

    override fun hashCode(): Int {
        return Objects.hash(storyId)
    }

    companion object {
        // --- الثوابت ---
        const val MEDIA_TYPE_IMAGE: String = "image"
        const val MEDIA_TYPE_VIDEO: String = "video"
    }
}